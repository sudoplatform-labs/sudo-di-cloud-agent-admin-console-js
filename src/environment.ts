export const uiConfig = '/uiconfig.json';
export const acapyConfig = '/acapy.json';

export interface EnvironmentInfo {
  acapyAdminUri: string; // REST API For Cloud Agent
  acapyAdminKey: string; // REST API authorisation key
  featureFlags: string[]; // Enabled/Disbled features for this customer
}
