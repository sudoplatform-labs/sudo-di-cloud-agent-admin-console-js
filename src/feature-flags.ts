/**
 * Feature flags are set during ACC deployment via runtime deployer variables.
 * E.g. RD_FF_EXAMPLE_FEATURE => FeatureFlag.EXAMPLE_FEATURE
 */
export enum FeatureFlag {
  EXAMPLE_FEATURE,
}

export function getFeatureFlags(featureFlagStrings: string[]): FeatureFlag[] {
  return featureFlagStrings
    .map((ffs) => {
      return FeatureFlag[ffs as keyof typeof FeatureFlag];
    })
    .filter((flag) => flag !== undefined)
    .sort();
}
