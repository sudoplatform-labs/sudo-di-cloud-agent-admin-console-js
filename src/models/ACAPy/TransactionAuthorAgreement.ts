// We use snake_case in the models modules for protocol related
// fields because the Aries RFCs
// define the protocol using that schema.  Originally I attempted to
// convert to camel case here to isolate that peculiarity, however that
// resulted in confusion when comparing to on-the-wire data. In addition
// because of an issue with OpenAPI code generation of typescript-fetch,
// if we don't use 'orginal' for the names output the generated
// api doesn't match the json objects returned in responses.

/*
 * This module contains all routines and definitions required to
 * interface with the Cloud Agent service for activities related to
 * Transaction Author Agreement processing.
 */
import { CloudAgentAPI } from '../../containers/App/AppContext';
import { reportCloudAgentError } from '../../utils/errorlog';
import { TAAInfo, TAAResult } from '@sudoplatform-labs/sudo-di-cloud-agent';

export type TAAAcceptParams = {
  // The way the TAA was displayed and acknowledged.
  // Must be one of the key values returned in the aml_record.aml[]
  mechanism: string;
  // The EXACT string displayed to the user and accepted.
  // This will be used in signing and must match the ledgers TAA
  text: string;
  // The taa_record.version value from the fetchLedgerTaa() result.
  version?: string;
};

export async function fetchLedgerTaa(agent: CloudAgentAPI): Promise<TAAInfo> {
  try {
    const ledgerTaa: TAAResult = await agent.ledger.ledgerTaaGet();

    return ledgerTaa.result ?? {};
  } catch (error) {
    throw await reportCloudAgentError(
      `Current Transaction Author Agreement NOT RETURNED by Ledger`,
      error,
    );
  }
}

export async function acceptLedgerTaa(
  agent: CloudAgentAPI,
  params: TAAAcceptParams,
): Promise<void> {
  try {
    await agent.ledger.ledgerTaaAcceptPost({
      body: {
        mechanism: params.mechanism,
        text: params.text,
        version: params.version,
      },
    });
  } catch (error) {
    throw await reportCloudAgentError(
      `Transaction Author Agreement NOT ACCEPTED by Ledger`,
      error,
    );
  }
}
