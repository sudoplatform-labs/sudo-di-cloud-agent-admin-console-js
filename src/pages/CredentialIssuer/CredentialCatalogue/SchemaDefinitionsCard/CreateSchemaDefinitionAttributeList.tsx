import Table, { ColumnProps, TableProps } from 'antd/lib/table';
import React from 'react';
import styled from 'styled-components';
import { consoleTableMixin, DangerLink } from '../../../../components/table';

export type SchemaAttribute = {
  name: string;
};

const AttributesTable = Table as React.FC<TableProps<SchemaAttribute>>;

const StyledConsoleTable = styled(AttributesTable)`
  ${consoleTableMixin}
  & tr > td,
  tr > th {
    padding: 16px 24px;
  }
  ul.ant-pagination.ant-table-pagination {
    margin-right: 24px;
  }
`;

interface Props {
  dataSource: SchemaAttribute[];
  loading?: boolean;
  onDelete: (attribute: SchemaAttribute) => void;
}

export const CreateSchemaDefinitionAttributeList: React.FC<Props> = (props) => {
  const { dataSource, loading, onDelete: doRemove } = props;

  const makeColumns = (opts: {
    onRemove: (attribute: SchemaAttribute) => void;
  }): ColumnProps<SchemaAttribute>[] => {
    return [
      {
        title: 'Attribute Name',
        dataIndex: 'name',
      },
      {
        title: <span style={{ visibility: 'hidden' }}>{'Remove'}</span>,
        dataIndex: 'id',
        align: 'right',
        render(_, attributeInfo) {
          return (
            <DangerLink
              onClick={() => opts.onRemove(attributeInfo)}
              className="CreateSchemaDeginitionAttributeList__removeAttributeButton">
              Remove
            </DangerLink>
          );
        },
        sorter: (a, b) => ('' + a.name).localeCompare(b.name),
        sortDirections: ['descend', 'ascend', 'descend'],
        defaultSortOrder: 'ascend',
      },
    ];
  };

  const columns = makeColumns({
    onRemove: doRemove,
  });

  return (
    <StyledConsoleTable
      id="CreateSchemaDefinitionAttributeList"
      loading={loading}
      columns={columns}
      dataSource={dataSource}
      rowKey={(record) => record.name}
      pagination={{ pageSize: 5 }}
    />
  );
};
