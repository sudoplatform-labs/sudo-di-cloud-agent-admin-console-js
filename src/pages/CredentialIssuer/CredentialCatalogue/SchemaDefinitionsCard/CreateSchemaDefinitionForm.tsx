import React, { useRef, useState } from 'react';
import { Form, Input, message, Alert, Button } from 'antd';
import { Store } from 'rc-field-form/lib/interface';
import { useAsyncFn } from 'react-use';
import styled from 'styled-components';
import { CloudAgentAPI } from '../../../../containers/App';
import {
  createSchemaDefinition,
  SchemaDefinitionCreateParams,
} from '../../../../models/ACAPy/SchemaDefinitions';
import {
  CreateSchemaDefinitionAttributeList,
  SchemaAttribute,
} from './CreateSchemaDefinitionAttributeList';
import { modalTAAAcceptance } from '../../../../components/Form';
import { RuleObject } from 'antd/lib/form';
import { StoreValue } from 'antd/lib/form/interface';

const FooterContainer = styled.div`
  display: flex;
  justify-content: flex-end;
  margin: 0 -24px -8px;
  padding: 18px 24px 0;
`;

const StyledButton = styled(Button)`
  min-width: 112px;
`;

const SubmitButton = styled(StyledButton)`
  margin-left: 12px;
`;

interface Props {
  onSuccess: () => void;
  onCancel: () => void;
  cloudAgentAPIs: CloudAgentAPI;
}

export const CreateSchemaDefinitionForm: React.FC<Props> = (props) => {
  const { onCancel, onSuccess, cloudAgentAPIs } = props;
  const [form] = Form.useForm();
  // Track the current set of schema attributes to adjust
  // the display as new ones are added
  const [currentAttributes, setCurrentAttributes] = useState<SchemaAttribute[]>(
    [],
  );

  const [submitState, handleSubmit] = useAsyncFn(
    async (values?: Store) => {
      // On initial entry we don't want to do anything
      if (!values) {
        return undefined;
      }

      const definitionParams: SchemaDefinitionCreateParams = {
        name: values.schemaName,
        version: values.schemaVersion,
        attributes: currentAttributes.map((attribute) => attribute.name),
      };

      try {
        await modalTAAAcceptance(cloudAgentAPIs, async () => {
          await createSchemaDefinition(cloudAgentAPIs, definitionParams);
          message.success('Schema Definition Created!');
          form.resetFields();
          setCurrentAttributes([]);
          onSuccess();
        });
      } catch {
        message.error('Unable to write to public ledger, please try again');
      }
    },
    [cloudAgentAPIs, currentAttributes, form, onSuccess],
  );

  const deleteAttributeHandler = (attribute: SchemaAttribute): void => {
    const index = currentAttributes.indexOf(attribute);
    setCurrentAttributes(
      currentAttributes
        .slice(0, index)
        .concat(currentAttributes.slice(index + 1)),
    );
  };

  const attributeInput = useRef<Input>(null);

  const enterAttributeNameHandler = async (
    keyEvent: React.KeyboardEvent<HTMLInputElement>,
  ): Promise<void> => {
    keyEvent.preventDefault(); // Don't let the form submit on enter here
    const attributeName = keyEvent.currentTarget.value;
    try {
      setCurrentAttributes([...currentAttributes, { name: attributeName }]);
      form.resetFields(['attributes']);
      // This is bizzare but for some reason eslint complains if
      // you try and access the ref directly even if you put it
      // inside an if condition checking it and current aren't null !
      const node = attributeInput.current;
      if (node) {
        node.focus();
      }
    } catch (err) {
      // Do nothing on bad input
      message.error(err as Error);
    }
  };

  return (
    <Form
      id="CreateSchemaDefinitionForm"
      form={form}
      onFinish={handleSubmit}
      layout="vertical">
      <Form.Item
        name="schemaName"
        label="Schema Name"
        rules={[
          {
            required: true,
            validator: (rule: RuleObject, value: StoreValue): Promise<void> => {
              const validSchemaNameRegex = /^[a-zA-Z0-9_-]+$/;
              // Vaidate that the pattern matches a valid schema Id before attempting
              // to fetch it to avoid constant fetch attempts on every character entry.
              if (value && validSchemaNameRegex.test(value)) {
                return Promise.resolve(value);
              } else {
                return Promise.reject(
                  'Schema names can only contain the following characters a-zA-Z0-9_-',
                );
              }
            },
          },
        ]}>
        <Input
          type="text"
          placeholder="e.g. University_Transcript"
          onPressEnter={(e) => e.preventDefault()}
        />
      </Form.Item>
      <Form.Item
        name="schemaVersion"
        label="Version"
        rules={[
          {
            required: true,
            pattern: new RegExp('^[0-9]+.[0-9]$'),
            message: 'Please provide a valid Schema version number (e.g. 1.0).',
          },
        ]}>
        <Input
          type="text"
          placeholder="e.g. 1.0"
          onPressEnter={(e) => e.preventDefault()}
        />
      </Form.Item>
      <Form.Item
        name="attributes"
        label="Schema Attributes"
        rules={[
          {
            required: true,
            validator: (rule: RuleObject, value: StoreValue): Promise<void> => {
              const validSchemaNameRegex = /^[a-zA-Z0-9_-]+$/;
              // Vaidate that the pattern matches a valid schema Id before attempting
              // to fetch it to avoid constant fetch attempts on every character entry.
              if (value && !validSchemaNameRegex.test(value)) {
                return Promise.reject(
                  'Attribute names can only contain the following characters a-zA-Z0-9_-',
                );
              } else if (
                currentAttributes.find((attribute) => value === attribute.name)
              ) {
                return Promise.reject(
                  'Attribute names must be unique within a Schema',
                );
              } else {
                return Promise.resolve(value);
              }
            },
          },
        ]}>
        <Input
          name="attributeInput"
          type="text"
          placeholder="Type attribute name then <enter> to add"
          onPressEnter={enterAttributeNameHandler}
          ref={attributeInput}
        />
      </Form.Item>
      <CreateSchemaDefinitionAttributeList
        dataSource={currentAttributes}
        onDelete={deleteAttributeHandler}
      />
      {submitState.error && (
        <Alert
          type="error"
          message="There was an error creating your Schema Definition. Please try again."
          description={submitState.error.message}
        />
      )}
      <FooterContainer>
        <StyledButton
          id="CreateSchemaDefinitionForm__cancel-btn"
          onClick={() => {
            submitState.error = undefined;
            form.resetFields();
            setCurrentAttributes([]);
            onCancel();
          }}
          type="default"
          shape="round">
          Cancel
        </StyledButton>
        <SubmitButton
          id="CreateSchemaDefinitionForm__submit-btn"
          type="primary"
          htmlType="submit"
          loading={submitState.loading}
          shape="round">
          Create
        </SubmitButton>
      </FooterContainer>
    </Form>
  );
};
