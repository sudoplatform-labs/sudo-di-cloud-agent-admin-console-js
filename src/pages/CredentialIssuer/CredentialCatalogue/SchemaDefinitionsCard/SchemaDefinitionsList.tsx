import Table, { ColumnProps, TableProps } from 'antd/lib/table';
import React, { useState } from 'react';
import styled from 'styled-components';
import {
  consoleTableMixin,
  SearchState,
  getColumnSearchProps,
} from '../../../../components/table';
import { PlusSquareOutlined, MinusSquareOutlined } from '@ant-design/icons';
import { VStack } from '../../../../components/layout-stacks';
import { Heading } from '../../../../components/charts';
import { Schema } from '@sudoplatform-labs/sudo-di-cloud-agent';

const SchemaInfoTable = Table as React.FC<TableProps<Schema>>;

const StyledConsoleTable = styled(SchemaInfoTable)`
  ${consoleTableMixin}
  & tr > td,
  tr > th {
    padding: 16px 24px;
  }
  ul.ant-pagination.ant-table-pagination {
    margin-right: 24px;
  }
`;

interface Props {
  dataSource: Schema[];
  loading?: boolean;
}

export const SchemaDefinitionsList: React.FC<Props> = (props) => {
  const [searchState, setSearchState] = useState<SearchState>({
    searchText: '',
    searchedColumn: '',
  });

  const makeColumns = (): ColumnProps<Schema>[] => {
    return [
      {
        title: 'Name',
        dataIndex: 'name',
        ...getColumnSearchProps('name', searchState, setSearchState),
      },
      {
        title: 'Version',
        dataIndex: 'version',
        ...getColumnSearchProps('version', searchState, setSearchState),
      },
      {
        title: 'Sequence No.',
        dataIndex: 'seqNo',
        ...getColumnSearchProps('seqNo', searchState, setSearchState),
      },
    ];
  };

  const columns = makeColumns();

  return (
    <StyledConsoleTable
      id="SchemaDefinitionsList"
      loading={props.loading}
      columns={columns}
      dataSource={props.dataSource}
      rowKey={(record) => record.id ?? ''}
      expandable={{
        expandedRowRender: (record) => (
          <VStack>
            <Heading> Schema Identifier </Heading>
            <p> {record.id} </p>
            <Heading> Schema Attribute Names </Heading>
            <p> {record.attrNames?.join(', ')} </p>
            <Heading> Ledger Node Protocol Version </Heading>
            <p> {record.ver} </p>
          </VStack>
        ),
        expandIcon: ({ expanded, onExpand, record }) => {
          if (expanded)
            return <MinusSquareOutlined onClick={(e) => onExpand(record, e)} />;
          else
            return <PlusSquareOutlined onClick={(e) => onExpand(record, e)} />;
        },
      }}
    />
  );
};
