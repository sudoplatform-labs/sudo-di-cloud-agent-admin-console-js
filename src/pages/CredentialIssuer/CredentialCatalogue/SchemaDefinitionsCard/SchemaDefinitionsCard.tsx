import { InfoCircleOutlined } from '@ant-design/icons';
import { Button, Popover, Modal } from 'antd';
import React, { useCallback, useContext, useEffect, useState } from 'react';
import { useAsyncFn } from 'react-use';
import styled from 'styled-components';
import { ConsoleCard } from '../../../../components/ConsoleCard';
import { CardsCol, CardsRow } from '../../../../components/NavLayout';
import { theme } from '../../../../theme';
import { AppContext } from '../../../../containers/App';
import { fetchAllAgentSchemaDefinitionDetails } from '../../../../models/ACAPy/SchemaDefinitions';
import { CreateSchemaDefinitionForm } from './CreateSchemaDefinitionForm';
import { SchemaDefinitionsList } from './SchemaDefinitionsList';
import { Schema } from '@sudoplatform-labs/sudo-di-cloud-agent';

type ModalState = 'create-schema-open' | 'closed';

/**
 * Stylised hover information icon to explain the purpose of
 * Schema Definitions in the Decentralized Identification ecosystem.
 */
const StyledInfoCircleOutlined = styled(InfoCircleOutlined)`
  color: ${theme.colors.sudoBlue};
`;

export const SchemaDefinitionsIconPopover: React.FC = () => {
  const content = (
    <p>
      Verifiable Credential Schemas define the attributes
      <br /> for a bound Verifiable Credential Definition.
      <br /> An Issuer can use any Schema on the Public Ledger to
      <br /> create a Verifiable Credential Definition tied to
      <br /> their own Decentralized Identifier.
    </p>
  );
  return (
    <Popover
      id="SchemaDefinitionsCard__popover-dialog"
      title="Schema Definitions"
      trigger="hover"
      content={content}>
      <StyledInfoCircleOutlined id="SchemaDefinitionsCard__popover-icon" />
    </Popover>
  );
};

/**
 * The SchemaDefinitionsCard React component deals with writing
 * Verifiable Credential Schemas to the Decentralized Ledger.
 * These are both Publicly Visible and publicly re-usable to
 * create Verified Credential Definitions.
 */
export const SchemaDefinitionsCard: React.FC = () => {
  const { cloudAgentAPIs } = useContext(AppContext);
  const [modalState, setModalState] = useState<ModalState>('closed');

  const [
    { loading: infoLoading, value: schemas, error: agentFailed },
    getSchemaInfo,
  ] = useAsyncFn(
    async (): Promise<Schema[]> =>
      fetchAllAgentSchemaDefinitionDetails(cloudAgentAPIs),
    [cloudAgentAPIs],
  );

  useEffect(() => {
    getSchemaInfo();
  }, [getSchemaInfo, modalState]);

  const modalCancelHandler = useCallback(() => {
    setModalState('closed');
  }, []);

  const modalSuccessHandler = useCallback(() => {
    setModalState('closed');
  }, []);

  let schemaData;
  if (agentFailed) {
    schemaData = <div>Unable to connect to the Cloud Agent Service</div>;
  } else {
    schemaData = (
      <SchemaDefinitionsList dataSource={schemas ?? []} loading={infoLoading} />
    );
  }

  return (
    <CardsRow>
      <CardsCol span={24}>
        <ConsoleCard
          id="SchemaDefinitionsCard"
          title={
            <span>
              My Schema Definitions <SchemaDefinitionsIconPopover />
            </span>
          }
          extra={
            <Button
              id="SchemaDefinitionsCard__create-btn"
              type="primary"
              shape="round"
              onClick={() => setModalState('create-schema-open')}
              disabled={infoLoading}>
              Create Schema Definition
            </Button>
          }>
          {schemaData}
        </ConsoleCard>
      </CardsCol>
      <Modal
        title="Create Schema Definition"
        visible={modalState === 'create-schema-open'}
        onCancel={modalCancelHandler}
        footer={null}>
        <CreateSchemaDefinitionForm
          onCancel={modalCancelHandler}
          onSuccess={modalSuccessHandler}
          cloudAgentAPIs={cloudAgentAPIs}
        />
      </Modal>
    </CardsRow>
  );
};
