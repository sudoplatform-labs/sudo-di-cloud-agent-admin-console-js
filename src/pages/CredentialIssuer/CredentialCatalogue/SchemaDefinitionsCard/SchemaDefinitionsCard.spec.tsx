import React from 'react';
import { SchemaDefinitionsCard } from './SchemaDefinitionsCard';
import { acwMount } from '../../../../test/enzyme';
import * as modelSchemaDefinitions from '../../../../models/ACAPy/SchemaDefinitions';
import * as modelTAA from '../../../../models/ACAPy/TransactionAuthorAgreement';
import { SchemaDefinitionsList } from './SchemaDefinitionsList';
import {
  AppState,
  AppContext,
  CloudAgentAPI,
} from '../../../../containers/App/AppContext';
import { EnvironmentInfo } from '../../../../environment';
import * as FeatureFlags from '../../../../feature-flags';
import {
  WalletApi,
  LedgerApi,
  ConnectionApi,
  IntroductionApi,
  TrustpingApi,
  SchemaApi,
  CredentialDefinitionApi,
  IssueCredentialApi,
  RevocationApi,
  CredentialsApi,
  PresentProofApi,
  Schema,
  TAAInfo,
} from '@sudoplatform-labs/sudo-di-cloud-agent';
import { Button, Modal } from 'antd';
import { act } from 'react-dom/test-utils';
import { changeFormInput, submitForm } from '../../../../test/form';

const mockEnvironment: EnvironmentInfo = {
  acapyAdminUri: 'localhost:3000/',
  acapyAdminKey: 'testkey',
  featureFlags: ['EXAMPLE_FEATURE'],
};

const mockContext: AppState = {
  environment: mockEnvironment,
  featureFlags: [FeatureFlags.FeatureFlag.EXAMPLE_FEATURE],
  cloudAgentAPIs: {
    wallet: new WalletApi(),
    ledger: new LedgerApi(),
    introductions: new IntroductionApi(),
    connections: new ConnectionApi(),
    ping: new TrustpingApi(),
    defineSchemas: new SchemaApi(),
    defineCredentials: new CredentialDefinitionApi(),
    issueCredentials: new IssueCredentialApi(),
    revocations: new RevocationApi(),
    credentials: new CredentialsApi(),
    proofs: new PresentProofApi(),
    httpOptionOverrides: {
      httpPostOptionOverrides: {},
    },
  },
};

const mockSchemaDefinitions: Schema[] = [
  {
    id: 'TEST_SCHEMA_ID_0',
    name: 'TEST_SCHEMA_NAME_0',
    version: '0.1',
    seqNo: 5,
    attrNames: ['TEST_ATTRIBUTE_0', 'TEST_ATTRIBUTE_1'],
    ver: 'TEST_LEDGER_NODE_VERSION',
  },
];

const fetchAllAgentSchemaDefinitionDetailsSpy = jest
  .spyOn(modelSchemaDefinitions, 'fetchAllAgentSchemaDefinitionDetails')
  .mockResolvedValue(mockSchemaDefinitions);

const createSchemaDefinitionSpy = jest
  .spyOn(modelSchemaDefinitions, 'createSchemaDefinition')
  .mockImplementation(
    async (
      cloudAgentAPI: CloudAgentAPI,
      schemaDefInfo: modelSchemaDefinitions.SchemaDefinitionCreateParams,
    ) => {
      const newSchemaDef: Schema = {
        id: 'TEST_SCHEMA_ID_1',
        name: schemaDefInfo.name,
        version: schemaDefInfo.version,
        seqNo: 9,
        attrNames: schemaDefInfo.attributes,
        ver: 'TEST_LEDGER_NODE_VERSION',
      };
      mockSchemaDefinitions.push(newSchemaDef);
    },
  );

const mockTaa: TAAInfo = {
  taa_required: false,
};

const fetchLedgerTaaSpy = jest
  .spyOn(modelTAA, 'fetchLedgerTaa')
  .mockResolvedValue(mockTaa);

describe('SchemaDefinitionsCard', () => {
  it('should get schema list', async () => {
    const component = await acwMount(
      <AppContext.Provider value={mockContext}>
        <SchemaDefinitionsCard />
      </AppContext.Provider>,
    );
    const schemaDefinitionsList = component.find(SchemaDefinitionsList);
    expect(fetchAllAgentSchemaDefinitionDetailsSpy).toBeCalled();
    expect(schemaDefinitionsList.prop('dataSource')).toBe(
      mockSchemaDefinitions,
    );
  });

  it('should expand a schema definition row', async () => {
    const component = await acwMount(
      <AppContext.Provider value={mockContext}>
        <SchemaDefinitionsCard />
      </AppContext.Provider>,
    );

    const schemaDefinitionRow = component
      .find(SchemaDefinitionsList)
      .findWhere((node) => {
        return node.prop('rowKey') === mockSchemaDefinitions[0].id;
      });

    await act(async () => {
      schemaDefinitionRow.find('AntdIcon').simulate('click');
    });

    component.update();
    expect(schemaDefinitionRow).toMatchSnapshot();
    // JC*** how do you check expansion happened ???
  });

  it('should open, add schema definition modal', async () => {
    const component = await acwMount(
      <AppContext.Provider value={mockContext}>
        <SchemaDefinitionsCard />
      </AppContext.Provider>,
    );
    const addButton = component
      .find(Button)
      .filterWhere(
        (node) => node.prop('id') === 'SchemaDefinitionsCard__create-btn',
      );

    await act(async () => {
      addButton.simulate('click');
    });

    component.update();
    const modal = component.find(Modal);
    expect(modal.prop('visible')).toBe(true);
  });

  it('should close, add schema definition modal', async () => {
    const component = await acwMount(
      <AppContext.Provider value={mockContext}>
        <SchemaDefinitionsCard />
      </AppContext.Provider>,
    );
    const addButton = component
      .find(Button)
      .filterWhere(
        (node) => node.prop('id') === 'SchemaDefinitionsCard__create-btn',
      );
    await act(async () => {
      addButton.simulate('click');
    });
    component.update();
    const modal = component.find(Modal);
    expect(modal.prop('visible')).toBe(true);

    const cancelButton = modal
      .find(Button)
      .filterWhere(
        (node) => node.prop('id') === 'CreateSchemaDefinitionForm__cancel-btn',
      );
    await act(async () => {
      cancelButton.simulate('click');
    });
    component.update();
    const updatedModal = component.find(Modal);
    expect(updatedModal.prop('visible')).toBe(false);
  });

  it('should add a new schema definition and display', async () => {
    const component = await acwMount(
      <AppContext.Provider value={mockContext}>
        <SchemaDefinitionsCard />
      </AppContext.Provider>,
    );
    const addButton = component
      .find(Button)
      .filterWhere(
        (node) => node.prop('id') === 'SchemaDefinitionsCard__create-btn',
      );
    await act(async () => {
      addButton.simulate('click');
    });
    component.update();
    const modal = component.find(Modal);
    expect(modal.prop('visible')).toBe(true);

    await act(async () => {
      await changeFormInput(modal, 'schemaName', 'TEST_SCHEMA_NAME_NEW');
      await changeFormInput(modal, 'schemaVersion', '0.1');
      await changeFormInput(
        modal,
        'attributes',
        'attribute_one attribute_two attribute_three',
      );
    });

    await submitForm(modal);
    component.update();
    const updatedModal = component.find(Modal);
    expect(fetchAllAgentSchemaDefinitionDetailsSpy).toBeCalledTimes(3);
    expect(fetchLedgerTaaSpy).toBeCalledTimes(1);
    expect(createSchemaDefinitionSpy).toBeCalledTimes(1);
    expect(updatedModal.prop('visible')).toBe(false);
    const schemaDefinitionsList = component.find(SchemaDefinitionsList);
    expect(schemaDefinitionsList.prop('dataSource')).toBe(
      mockSchemaDefinitions,
    );
  });
});
