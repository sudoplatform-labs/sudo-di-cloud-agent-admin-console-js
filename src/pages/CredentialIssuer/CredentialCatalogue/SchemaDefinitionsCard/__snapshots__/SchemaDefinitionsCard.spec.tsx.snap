// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemaDefinitionsCard should expand a schema definition row 1`] = `
<BodyRow
  cellComponent="td"
  childrenColumnName="children"
  expandedKeys={Set {}}
  getRowKey={[Function]}
  index={0}
  key="TEST_SCHEMA_ID_0"
  record={
    Object {
      "attrNames": Array [
        "TEST_ATTRIBUTE_0",
        "TEST_ATTRIBUTE_1",
      ],
      "id": "TEST_SCHEMA_ID_0",
      "name": "TEST_SCHEMA_NAME_0",
      "seqNo": 5,
      "ver": "TEST_LEDGER_NODE_VERSION",
      "version": "0.1",
    }
  }
  recordKey="TEST_SCHEMA_ID_0"
  rowComponent="tr"
  rowKey="TEST_SCHEMA_ID_0"
>
  <tr
    className="ant-table-row ant-table-row-level-0"
    data-row-key="TEST_SCHEMA_ID_0"
    onClick={[Function]}
    style={Object {}}
  >
    <Cell
      className="ant-table-row-expand-icon-cell"
      component="td"
      firstFixLeft={false}
      firstFixRight={false}
      index={0}
      key="RC_TABLE_KEY"
      lastFixLeft={false}
      lastFixRight={false}
      prefixCls="ant-table"
      record={
        Object {
          "attrNames": Array [
            "TEST_ATTRIBUTE_0",
            "TEST_ATTRIBUTE_1",
          ],
          "id": "TEST_SCHEMA_ID_0",
          "name": "TEST_SCHEMA_NAME_0",
          "seqNo": 5,
          "ver": "TEST_LEDGER_NODE_VERSION",
          "version": "0.1",
        }
      }
      render={[Function]}
    >
      <td
        className="ant-table-cell ant-table-row-expand-icon-cell"
        colSpan={null}
        rowSpan={null}
        style={Object {}}
      >
        <ForwardRef(PlusSquareOutlined)
          onClick={[Function]}
        >
          <AntdIcon
            icon={
              Object {
                "icon": Object {
                  "attrs": Object {
                    "focusable": "false",
                    "viewBox": "64 64 896 896",
                  },
                  "children": Array [
                    Object {
                      "attrs": Object {
                        "d": "M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z",
                      },
                      "tag": "path",
                    },
                    Object {
                      "attrs": Object {
                        "d": "M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z",
                      },
                      "tag": "path",
                    },
                  ],
                  "tag": "svg",
                },
                "name": "plus-square",
                "theme": "outlined",
              }
            }
            onClick={[Function]}
          >
            <span
              aria-label="plus-square"
              className="anticon anticon-plus-square"
              onClick={[Function]}
              role="img"
              tabIndex={-1}
            >
              <IconReact
                icon={
                  Object {
                    "icon": Object {
                      "attrs": Object {
                        "focusable": "false",
                        "viewBox": "64 64 896 896",
                      },
                      "children": Array [
                        Object {
                          "attrs": Object {
                            "d": "M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z",
                          },
                          "tag": "path",
                        },
                        Object {
                          "attrs": Object {
                            "d": "M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z",
                          },
                          "tag": "path",
                        },
                      ],
                      "tag": "svg",
                    },
                    "name": "plus-square",
                    "theme": "outlined",
                  }
                }
              >
                <svg
                  aria-hidden="true"
                  data-icon="plus-square"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  key="svg-plus-square"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
                    key="svg-plus-square-svg-0"
                  />
                  <path
                    d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
                    key="svg-plus-square-svg-1"
                  />
                </svg>
              </IconReact>
            </span>
          </AntdIcon>
        </ForwardRef(PlusSquareOutlined)>
      </td>
    </Cell>
    <Cell
      component="td"
      dataIndex="name"
      firstFixLeft={false}
      firstFixRight={false}
      index={0}
      key="name"
      lastFixLeft={false}
      lastFixRight={false}
      prefixCls="ant-table"
      record={
        Object {
          "attrNames": Array [
            "TEST_ATTRIBUTE_0",
            "TEST_ATTRIBUTE_1",
          ],
          "id": "TEST_SCHEMA_ID_0",
          "name": "TEST_SCHEMA_NAME_0",
          "seqNo": 5,
          "ver": "TEST_LEDGER_NODE_VERSION",
          "version": "0.1",
        }
      }
    >
      <td
        className="ant-table-cell"
        colSpan={null}
        rowSpan={null}
        style={Object {}}
      >
        TEST_SCHEMA_NAME_0
      </td>
    </Cell>
    <Cell
      component="td"
      dataIndex="version"
      firstFixLeft={false}
      firstFixRight={false}
      index={0}
      key="version"
      lastFixLeft={false}
      lastFixRight={false}
      prefixCls="ant-table"
      record={
        Object {
          "attrNames": Array [
            "TEST_ATTRIBUTE_0",
            "TEST_ATTRIBUTE_1",
          ],
          "id": "TEST_SCHEMA_ID_0",
          "name": "TEST_SCHEMA_NAME_0",
          "seqNo": 5,
          "ver": "TEST_LEDGER_NODE_VERSION",
          "version": "0.1",
        }
      }
    >
      <td
        className="ant-table-cell"
        colSpan={null}
        rowSpan={null}
        style={Object {}}
      >
        0.1
      </td>
    </Cell>
    <Cell
      component="td"
      dataIndex="seqNo"
      firstFixLeft={false}
      firstFixRight={false}
      index={0}
      key="seqNo"
      lastFixLeft={false}
      lastFixRight={false}
      prefixCls="ant-table"
      record={
        Object {
          "attrNames": Array [
            "TEST_ATTRIBUTE_0",
            "TEST_ATTRIBUTE_1",
          ],
          "id": "TEST_SCHEMA_ID_0",
          "name": "TEST_SCHEMA_NAME_0",
          "seqNo": 5,
          "ver": "TEST_LEDGER_NODE_VERSION",
          "version": "0.1",
        }
      }
    >
      <td
        className="ant-table-cell"
        colSpan={null}
        rowSpan={null}
        style={Object {}}
      >
        5
      </td>
    </Cell>
  </tr>
</BodyRow>
`;
